mlos_bench/mlos_bench/environments/__init__.py

```diff
@@ -63,14 +101,250 @@
+
+First, let's check the tunable parameters:
+
+>>> assert env.tunable_params.get_param_values() == {
+...    "dummy_param": "dummy",
+...    "dummy_param_int": 0,
+...    "dummy_param_float": 0.5,
+...    "dummy_param3": 0.0
+... }
+
+We can see the tunables from ``dummy_params_group1`` and ``dummy_params_group2`` groups specified
+via ``$tunables_ref1``, as well as the tunables from ``dummy_params_group3`` that we specified
+directly in the Environment config. All tunables are initialized to their default values.
+
+Now let's see how the variable propagation works.
+
+>>> env.const_args["const_arg_1"]
+'Default value of const_arg_1'
+
+``const_arg_1`` has the value we have assigned in the ``"const_args"`` section of the
+Environment config. No surprises here.
 
-Since :py:mod:`json configs <mlos_bench.config>` also support ``$variable``
-substitution in the values using the `globals` mechanism, this setting can used to
-dynamically change the set of active TunableGroups for a given Experiment using only
-`globals`, allowing for configs to be more modular and composable.
+>>> env.const_args["const_arg_from_globals_1"]
+'Substituted from globals - 1'
+>>> env.const_args["const_arg_from_globals_2"]
+'Substituted from globals - 2'
+
+``const_arg_from_globals_1`` and ``const_arg_from_globals_2`` were declared in the Environment's
+``const_args`` and ``required_args`` sections, respectively. Their values were overridden by the
+values from the global config.
+
+>>> env.const_args["const_arg_from_cli_1"]
+'Substituted from CLI - 1'
+>>> env.const_args["const_arg_from_cli_2"]
+'Substituted from CLI - 2'
+
+Likewise, ``const_arg_from_cli_1`` and ``const_arg_from_cli_2`` got their values from the
+command line. Note that for ``const_arg_from_cli_1`` the value from the command line takes
+precedence over the values specified in the Environment's ``const_args`` section **and** the one
+in the global config.
+
+Now let's set up the environment and see how the constant and tunable parameters get combined.
+We'll also assign some non-default values to the tunables, as the optimizer would do on each
+trial.
+
+>>> env.tunable_params["dummy_param_int"] = 99
+>>> env.tunable_params["dummy_param3"] = 0.999
+>>> with env:
+...     assert env.setup(env.tunable_params)
+...     assert env.parameters == {
+...         "const_arg_1": "Default value of const_arg_1",
+...         "const_arg_from_globals_1": "Substituted from globals - 1",
+...         "const_arg_from_globals_2": "Substituted from globals - 2",
+...         "const_arg_from_cli_1": "Substituted from CLI - 1",
+...         "const_arg_from_cli_2": "Substituted from CLI - 2",
+...         "trial_id": 1,
+...         "trial_runner_id": 1,
+...         "experiment_id": "test_experiment",
+...         "dummy_param": "dummy",
+...         "dummy_param_int": 99,
+...         "dummy_param_float": 0.5,
+...         "dummy_param3": 0.999
+...     }
```

Comment:
In general I like the layout you have of this example much better than mine.

I think in a future PR we can tackle documenting how these variables interact with `run` phase commands for `ScriptEnvs`.

However, I think it still might make sense to document how parent env propagation works since it has some impact on whether `const_args` and `required_args` are propagated thru to child envs, right?

Maybe we can do that in the `CompositeEnv` documentation directly and just reference it here?

New PR for that too?

What do you think?

Reply:
> In general I like the layout you have of this example much better than mine.
> 
> I think in a future PR we can tackle documenting how these variables interact with `run` phase commands for `ScriptEnvs`.

https://github.com/microsoft/MLOS/issues/965

> However, I think it still might make sense to document how parent env propagation works since it has some impact on whether `const_args` and `required_args` are propagated thru to child envs, right?
> 
> Maybe we can do that in the `CompositeEnv` documentation directly and just reference it here?

https://github.com/microsoft/MLOS/issues/964

> New PR for that too?
> 
> What do you think?